waku/v2/node/storage/message/waku_message_store.nim

```diff
@@ -1,425 +1,265 @@
-  var oldestReceiverTimestamp: int64
-  proc handler(s: ptr sqlite3_stmt) =
-    oldestReceiverTimestamp = column_timestamp(s, 0)
-  let query = "SELECT MIN(receiverTimestamp) FROM " & TABLE_TITLE;
-  let queryRes = db.query(query, handler)
-  if queryRes.isErr:
-    return err("failed to get the oldest receiver timestamp from the DB")
-  ok(oldestReceiverTimestamp)
-
-proc deleteOldestTime(db: WakuMessageStore): MessageStoreResult[void] =
-  # delete if there are messages in the DB that exceed the retention time by 10% and more (batch delete for efficiency)
-  let retentionTimestamp = getNanosecondTime(getTime().toUnixFloat()) - db.retentionTime.nanoseconds
-  let thresholdTimestamp = retentionTimestamp - db.retentionTime.nanoseconds div 10
-  if thresholdTimestamp <= db.oldestReceiverTimestamp or db.oldestReceiverTimestamp == 0: return ok()
-
-  var deleteQuery = "DELETE FROM " & TABLE_TITLE & " " &
-                          "WHERE receiverTimestamp < " & $retentionTimestamp
-
-  let res = db.database.query(deleteQuery, proc(s: ptr sqlite3_stmt) = discard)
-  if res.isErr:
-    return err(res.error)
-
-  info "Messages exceeding retention time deleted from DB. ", retentionTime=db.retentionTime
-
-  db.oldestReceiverTimestamp = db.database.getOldestDbReceiverTimestamp().expect("DB query works")
 
-  ok()
+proc calculateTotalCapacity(capacity: int, overflow: float): int {.inline.} =
+  int(float(capacity) * overflow)
+
+proc calculateOverflowWindow(capacity: int, overflow: float): int {.inline.} =
+  int(float(capacity) * (overflow - 1))
+  
+proc calculateDeleteWindow(capacity: int, overflow: float): int {.inline.} =
+  calculateOverflowWindow(capacity, overflow) div 2
+
+
+### Store implementation
 
-proc deleteOldest(db: WakuMessageStore): MessageStoreResult[void] =
-  var deleteQuery = "DELETE FROM " & TABLE_TITLE & " " &
-                          "WHERE id NOT IN " &
-                              "(SELECT id FROM " & TABLE_TITLE & " " &
-                                "ORDER BY receiverTimestamp DESC " &
-                                "LIMIT " & $(db.storeCapacity + db.deleteWindow) & ")"
-  let res = db.database.query(deleteQuery, proc(s: ptr sqlite3_stmt) = discard)
-  if res.isErr:
-    return err(res.error)
-  db.numMessages = db.storeCapacity + db.deleteWindow # sqlite3 DELETE does not return the number of deleted rows; Ideally we would subtract the number of actually deleted messages. We could run a separate COUNT.
-
-  info "Oldest messages deleted from DB due to overflow.", storeCapacity=db.storeCapacity, maxStore=db.storeMaxLoad, deleteWindow=db.deleteWindow
-  when defined(debug):
-    let numMessages = messageCount(db.database).get() # requires another SELECT query, so only run in debug mode
-    debug "Number of messages left after delete operation.", messagesLeft=numMessages
+proc deleteMessagesExceedingRetentionTime(s: WakuMessageStore): MessageStoreResult[void] =
```

Comment:
IMHO we should not have this `isSqliteOnly` flag. The name is not descriptive enough in the context of this implementation and for me was a headache to understand how to refactor it 😅 

I would prefer having two different implementations of the `MessageStore` object that implement two different behaviors if the differences are notable or keep that flag as an SQLite message store configuration flag with a much better name.

I prefer the "two different store types" even if it means that we are duplicating some code in this case. But I kept what we had. I didn't want to diverge too much.

Reply:
Regarding where to update the state... Probably it's just my personal opinion/preference:
 
When I am modifying the state of an object I always try to update it at the "top-level functions" of that object (mainly, the methods that form the public API of that object). I try to avoid changing the state in helper methods because it makes it harder to follow the execution and data/state manipulation flow.

Said that, I am open to any suggestions 😁 If necessary we can iterate again and improve this implementation